### PR TITLE
No canvas

### PR DESCRIPTION
**Pull request by davidpiuva**
There is no more canvas.

After this PR, I will look for redundancy in the code that can be removed.
This PR is just moving things with the minimum possible modification.

Builds on https://github.com/magic-lang/ooc-kean/pull/1726.

### CODE REVIEW

**Comment by davidpiuva** [comment]
@emilwestergren Peer review?

**Comment by davidpiuva** [comment]
Same leaks as usual.
```
==11925== Memcheck, a memory error detector
==11925== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==11925== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==11925== Command: ./Tests -n
==11925== Parent PID: 11924
==11925== 
==11925== 
==11925== HEAP SUMMARY:
==11925==     in use at exit: 161,286 bytes in 464 blocks
==11925==   total heap usage: 17,247,945 allocs, 17,247,481 frees, 1,240,242,983 bytes allocated
==11925== 
==11925== 0 bytes in 1 blocks are definitely lost in loss record 1 of 193
==11925==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11925==    by 0x4C4C60: GraphicBuffer__GraphicBuffer___load__ (GraphicBuffer.ooc:105)
==11925==    by 0x4C5175: GraphicBuffer_load (GraphicBuffer.ooc:308)
==11925==    by 0x4C4214: AndroidContext_load (AndroidContext.ooc:383)
==11925==    by 0x4C298B: OpenGLWindow_load (OpenGLWindow.ooc:144)
==11925==    by 0x4840A1: test_draw_RasterCanvasTest_load (RasterCanvasTest.ooc:360)
==11925==    by 0x4808F2: ___test_Tests_load (Tests.c:108)
==11925==    by 0x4809D5: main (Tests.c:134)
==11925== 
==11925== 32 bytes in 1 blocks are definitely lost in loss record 93 of 193
==11925==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11925==    by 0x53278B: types__Class_alloc__class (types.ooc:61)
==11925==    by 0x5367F7: Dynlib__DynlibUnix_new (Dynlib.ooc:74)
==11925==    by 0x5365B6: Dynlib__Dynlib_load (Dynlib.ooc:20)
==11925==    by 0x548E5E: Backtrace__BacktraceHandler_init (Backtrace.ooc:88)
==11925==    by 0x549837: Backtrace__BacktraceHandler_new (Backtrace.ooc:81)
==11925==    by 0x549870: Backtrace__BacktraceHandler_get (Backtrace.ooc:181)
==11925==    by 0x52F470: Exception__Exception_addBacktrace (Exception.ooc:113)
==11925==    by 0x52F761: Exception__Exception_throw (Exception.ooc:137)
==11925==    by 0x530587: Exception__raise_withClass (Exception.ooc:85)
==11925==    by 0x4E6591: StbImage__StbImage_load (StbImage.ooc:22)
==11925==    by 0x4E4A3E: RasterRgba__RasterRgba_open (RasterRgba.ooc:152)
==11925==    by 0x49BCC0: test_draw_RasterRgbaTest____test_draw_RasterRgbaTest_test_draw_RasterRgbaTest_closure29 (RasterRgbaTest.ooc:21)
==11925==    by 0x4D2161: Fixture__Test_run (Fixture.ooc:196)
==11925==    by 0x4D05CD: Fixture__Fixture_run (Fixture.ooc:42)
==11925==    by 0x49B83E: test_draw_RasterRgbaTest_load (RasterRgbaTest.ooc:86)
==11925==    by 0x48099C: ___test_Tests_load (Tests.c:125)
==11925==    by 0x4809D5: main (Tests.c:134)
==11925== 
==11925== 32 bytes in 1 blocks are definitely lost in loss record 94 of 193
==11925==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11925==    by 0x53278B: types__Class_alloc__class (types.ooc:61)
==11925==    by 0x5367F7: Dynlib__DynlibUnix_new (Dynlib.ooc:74)
==11925==    by 0x5365B6: Dynlib__Dynlib_load (Dynlib.ooc:20)
==11925==    by 0x548E78: Backtrace__BacktraceHandler_init (Backtrace.ooc:88)
==11925==    by 0x549837: Backtrace__BacktraceHandler_new (Backtrace.ooc:81)
==11925==    by 0x549870: Backtrace__BacktraceHandler_get (Backtrace.ooc:181)
==11925==    by 0x52F470: Exception__Exception_addBacktrace (Exception.ooc:113)
==11925==    by 0x52F761: Exception__Exception_throw (Exception.ooc:137)
==11925==    by 0x530587: Exception__raise_withClass (Exception.ooc:85)
==11925==    by 0x4E6591: StbImage__StbImage_load (StbImage.ooc:22)
==11925==    by 0x4E4A3E: RasterRgba__RasterRgba_open (RasterRgba.ooc:152)
==11925==    by 0x49BCC0: test_draw_RasterRgbaTest____test_draw_RasterRgbaTest_test_draw_RasterRgbaTest_closure29 (RasterRgbaTest.ooc:21)
==11925==    by 0x4D2161: Fixture__Test_run (Fixture.ooc:196)
==11925==    by 0x4D05CD: Fixture__Fixture_run (Fixture.ooc:42)
==11925==    by 0x49B83E: test_draw_RasterRgbaTest_load (RasterRgbaTest.ooc:86)
==11925==    by 0x48099C: ___test_Tests_load (Tests.c:125)
==11925==    by 0x4809D5: main (Tests.c:134)
==11925== 
==11925== LEAK SUMMARY:
==11925==    definitely lost: 64 bytes in 3 blocks
==11925==    indirectly lost: 0 bytes in 0 blocks
==11925==      possibly lost: 0 bytes in 0 blocks
==11925==    still reachable: 4,717 bytes in 27 blocks
==11925==         suppressed: 156,505 bytes in 434 blocks
==11925== Reachable blocks (those to which a pointer was found) are not shown.
==11925== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==11925== 
==11925== For counts of detected and suppressed errors, rerun with: -v
==11925== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 80 from 80)
```

**Comment by davidpiuva** [comment]
PRInt has been stuck for hours now.

**Comment by marcusnaslund** [comment]
#1726 now merged, please rebase.

**Comment by davidpiuva** [comment]
Rebased

**Comment by marcusnaslund** [comment]
Thanks. I will unfortunately not have time to look at this today though.

**Comment by marcusnaslund** [comment]
Frozen awaiting release.
(But will be reviewed before that)

File: source/draw/Image.ooc [hunk 1]
```diff
@@ -38,22 +31,39 @@ Image: abstract class {
...
-	fill: func ~implicitCanvas (color: ColorRgba) { this canvas fill(color) }
+	drawPoint: virtual func ~white (position: FloatPoint2D) { this drawPoint(position, Pen new(ColorRgba white)) }
+	drawPoint: virtual func ~explicit (position: FloatPoint2D, pen: Pen) {
+		list := VectorList<FloatPoint2D> new()
+		list add(position)
+		this drawPoints(list, pen)
+		list free()
+	}
+	drawLine: virtual func ~white (start, end: FloatPoint2D) { this drawLine(start, end, Pen new(ColorRgba white)) }
+	drawLine: virtual func ~explicit (start, end: FloatPoint2D, pen: Pen) {
```
Comment (marcusnaslund) [inline source/draw/Image.ooc]: Can't you do `pen := Pen new(ColorRgba White)` in the arguments and avoid the extra declaration above? Same with many methods here.

File: source/draw/RasterRgba.ooc [hunk 2]
```diff
@@ -131,8 +125,6 @@ RasterRgba: class extends RasterPacked {
 		result swapRedBlue()
 		result
 	}
-	_createCanvas: override func -> Canvas { RasterRgbaCanvas new(this) }
-
 	operator [] (x, y: Int) -> ColorRgba { this isValidIn(x, y) ? ((this buffer pointer + y * this stride) as ColorRgba* + x)@ : ColorRgba new(0, 0, 0, 0) }
```
Comment (marcusnaslund) [inline source/draw/RasterRgba.ooc]: Not for this PR, but should `[]` really always call `isValidIn`? Seems pretty inefficient.

File: source/draw/gpu/opengl/OpenGLPacked.ooc [hunk 3]
```diff
@@ -30,25 +35,130 @@ OpenGLPacked: abstract class extends GpuImage {
...
 	init: func (=_backend, =_channels, context: OpenGLContext) {
 		super(this _backend size, context)
+		this _renderTarget = context _backend createFramebufferObject(this _backend as GLTexture, this size)
 	}
 	free: override func {
 		if (this recyclable)
 			this context recycle(this)
 		else {
 			this _backend free()
+			this _renderTarget free()
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/OpenGLPacked.ooc]: `(this _backend, this _renderTarget) free()`

**Comment by marcusnaslund** [comment]
Just a few comments, otherwise - really splendid work.

File: source/draw/RasterRgba.ooc (hunk 2 above)
Comment (davidpiuva) [inline source/draw/RasterRgba.ooc]: Local operations are supposed to read outside using constant offsets so this avoids having to reimplement safety in a less secure manner.

**Comment by davidpiuva** [comment]
Updated

**Pull request merged**